File: Algorithms_medium/90_Subsets II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        def dfs(u, t):
            ans.append(t[:])
            for i in range(u, len(nums)):
                if i != u and nums[i] == nums[i - 1]:
                    continue
                t.append(nums[i])
                dfs(i + 1, t)
                t.pop()

        ans = []
        nums.sort()
        dfs(0, [])
        return ans
'''
dfs搜索，与同类型的题目相似，不多做解释了
就是要记得整理数组，因为这是可重复数组
'''

File: Algorithms_C#/90_Subsets_II.cs
public class Solution {
    private IList<IList<int>> ans = new List<IList<int>>();
    private IList<int> t = new List<int>();
    private int[] nums;

    public IList<IList<int>> SubsetsWithDup(int[] nums) {
        Array.Sort(nums);
        this.nums = nums;
        dfs(0);
        return ans;
    }

    private void dfs(int i){
        if(i >= nums.Length){
            ans.Add(new List<int>(t));
            return;
        }
        t.Add(nums[i]);
        dfs(i+1);
        t.RemoveAt(t.Count - 1);
        while(i + 1 < nums.Length && nums[i+1] == nums[i]){
            ++i;
        }
        dfs(i+1);
    }
}